internal/sim/sim.go

```diff
@@ -220,54 +313,133 @@ func validateOp[T any](s *Simulator, o Op[T]) (op, error) {
 	}
 
 	return op{
-		t:    reflection.Type[T](),
-		name: o.Name,
-		gen:  o.Gen,
-		f:    o.Func,
+		t:          reflection.Type[T](),
+		name:       o.Name,
+		gen:        reflect.ValueOf(o.Gen),
+		f:          reflect.ValueOf(o.Func),
+		components: components,
 	}, nil
 }
 
-// GetIntf implements the Weavelet interface.
-func (s *Simulator) GetIntf(t reflect.Type) (any, error) {
-	reg, ok := s.regsByIntf[t]
-	if !ok {
-		return nil, fmt.Errorf("component %v not found", t)
-	}
-	call := func(method string, ctx context.Context, args []any, returns []any) error {
-		return s.call(reg, method, ctx, args, returns)
-	}
-	return reg.ReflectStubFn(call), nil
+// Simulate executes a single simulation.
+func (s *Simulator) Simulate(ctx context.Context) error {
+	s.group, s.ctx = errgroup.WithContext(ctx)
+	s.step()
+	return s.group.Wait()
 }
 
-// call executes a component method call against a random replica.
-func (s *Simulator) call(reg *codegen.Registration, method string, ctx context.Context, args []any, returns []any) error {
-	// Pick a replica to execute the call.
-	replicas := s.components[reg.Name]
+// step performs one step of a simulation.
+func (s *Simulator) step() {
 	s.mu.Lock()
-	i := s.rand.Intn(len(replicas))
-	s.mu.Unlock()
+	defer s.mu.Unlock()
 
-	// Convert the arguments to reflect.Values.
-	in := make([]reflect.Value, 1+len(args))
-	in[0] = reflect.ValueOf(ctx)
-	for i, arg := range args {
-		in[i+1] = reflect.ValueOf(arg)
+	if s.ctx.Err() != nil {
+		// The simulation has been cancelled.
+		return
 	}
 
-	// Call the method.
-	out := reflect.ValueOf(replicas[i]).MethodByName(method).Call(in)
+	// Compute the set of candidate steps.
+	const (
+		runOp = iota
+		deliverCall
+		deliverReply
+	)
+	var candidates []int
+	if s.numOps < s.opts.NumOps {
```

Comment:
This is an interesting algorithm for picking work items. I don't have anything better to suggest, just wanted to point out that it may converge fairly quickly to `s.opts.NumOps`. But correct me if I'm wrong, that is what we want since we'll start with smaller NumOps and then increase as needed?

Reply:
Ah, you're giving me too much credit. I just implemented the simplest possible thing I could think of! In the future, I hope to think harder about smart ways to interleave things.